### PR TITLE
add syncBN support for custom device

### PR DESCRIPTION
**Pull request by heidongxianhua**
Fixes #ISSUE_NUMBER
there are some hard checks for `cuda`, so I make optimize the check so that we can run it for other device.

### CODE REVIEW

File: torch/nn/modules/_functions.py [hunk 1]
```diff
@@ -70,7 +71,7 @@ def forward(self, input, weight, bias, running_mean, running_var, eps, momentum,
             # world_size * (2C + 1) -> world_size * C, world_size * C, world_size * 1
             mean_all, invstd_all, count_all = torch.split(combined, num_channels, dim=1)
 
-        if not torch.cuda.is_current_stream_capturing():
+        if not (torch.cuda.is_available() and torch.cuda.is_current_stream_capturing()):
```
Comment (albanD) [inline torch/nn/modules/_functions.py:74]: Why is this necessary?

File: torch/nn/modules/_functions.py [hunk 2]
```diff
@@ -49,7 +49,8 @@ def forward(self, input, weight, bias, running_mean, running_var, eps, momentum,
         # batch_norm_gather_stats_with_counts calculates global mean & invstd based on
         # all gathered mean, invstd and count.
         # for nccl backend, use the optimized version of all gather.
-        if process_group._get_backend_name() == 'nccl':
+        # The Gloo backend does not support `all_gather_into_tensor`.
+        if process_group._get_backend_name() != "gloo":
```
Comment (albanD) [inline torch/nn/modules/_functions.py:53]: cc @H-Huang does that look ok to you?

File: torch/nn/modules/_functions.py (hunk 1 above)
Comment (heidongxianhua) [reply torch/nn/modules/_functions.py:74]: > 
yeah, the func of `is_current_stream_capturing` is defined as the following, if we do not check the `cuda.is_available()`, it will raise error for other device except `cuda`. @albanD 
```
if not hasattr(torch._C, '_CudaStreamBase'):
    # Define dummy base classes
    torch._C.__dict__['_CUDAGraph'] = _dummy_type('_CUDAGraph')
    torch._C.__dict__['_graph_pool_handle'] = _dummy_type('_graph_pool_handle')
    torch._C.__dict__['_cuda_isCurrentStreamCapturing'] = _dummy_type('_cuda_isCurrentStreamCapturing')

from torch._C import _CUDAGraph  # noqa: F401
from torch._C import _graph_pool_handle
from torch._C import _cuda_isCurrentStreamCapturing


def is_current_stream_capturing():
    r"""
    Returns True if CUDA graph capture is underway on the current CUDA stream, False otherwise.

    If a CUDA context does not exist on the current device, returns False without initializing the context.
    """
    return _cuda_isCurrentStreamCapturing()
```

Comment (albanD) [reply torch/nn/modules/_functions.py:74]: Ho that is not nice, we should update that.
The change here sounds ok then.

File: torch/nn/modules/_functions.py (hunk 2 above)
Comment (heidongxianhua) [reply torch/nn/modules/_functions.py:53]: @H-Huang @fegin  sorry to bother could you have a look?

**Comment by heidongxianhua** [comment]
could you give an approve and all the tests are ok? @albanD

**Review by albanD — APPROVED** [review]
That sounds pretty safe to me so tentatively accepting!
If @H-Huang says it's not ok, we can revert!

File: torch/nn/modules/_functions.py (hunk 2 above)
Comment (heidongxianhua) [reply torch/nn/modules/_functions.py:53]: > This would fail for backends like MPI (or other 3rd party backends) are not gloo but which also dont support `all_gather_into_tensor`. Which backend are you targeting for this change and is there urgency behind this feature?
> 
> A better way would be to introduce collective fallbacks which try a collective on a backend and if it is not available then fallback to another collective. We want to move away from hard checks for backends in our code to be more backend agnostic. However this feature is not available and still involves discussion/design from the distributed team

yeah I will make a change,and we are custom device with privateuse1 backend, it is a cuda-like device, we also want to support syncBN and this optimize.
